### PR TITLE
New Tutorial - moveit_py jupyter notebook tutorial

### PR DESCRIPTION
**Pull request by peterdavidfagan**
### Description

**This pr is still of status WIP**

This pull request adds a tutorial for leveraging moveit_py within a jupyter notebook. A video tutorial still needs to be recorded and added.

### CODE REVIEW

**Comment by peterdavidfagan** [comment]
@sea-bass this pr should be ready to be tested again if you have the capacity for this.

**Comment by peterdavidfagan** [comment]
@henningkayser would it be possible to activate CI for this pr? It would be useful for checking if I need to amend anything, I have already installed and run pre-commit locally so I think the changes should pass CI.

File: doc/examples/jupyter_notebook_prototyping/launch/jupyter_notebook_prototyping.launch.py [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+        MoveItConfigsBuilder(
+            robot_name="panda", package_name="moveit_resources_panda_moveit_config"
+        )
+        .robot_description(file_path="config/panda.urdf.xacro")
+        .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
+        .moveit_cpp(
+            file_path=os.path.join(
+                get_package_share_directory("moveit2_tutorials"),
+                "/config/jupyter_notebook_prototyping.yaml",
+            )
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/launch/jupyter_notebook_prototyping.launch.py:23]:
```suggestion
            file_path=os.path.join(
                get_package_share_directory("moveit2_tutorials"),
                "config", "jupyter_notebook_prototyping.yaml",
            )
```

File: doc/examples/jupyter_notebook_prototyping/launch/jupyter_notebook_prototyping.launch.py [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+        )
+        .to_moveit_configs()
+    )
+
+    rviz_config_file = os.path.join(
+        get_package_share_directory("moveit2_tutorials"),
+        "/config/jupyter_notebook_prototyping.rviz",
+    )
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/launch/jupyter_notebook_prototyping.launch.py:31]:
```suggestion
    rviz_config_file = os.path.join(
        get_package_share_directory("moveit2_tutorials"),
        "config", "jupyter_notebook_prototyping.rviz",
    )
```

File: doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+We can then start the teleoperation node: ::
+
+        from moveit.servo_client.devices.ps4_dualshock import PS4DualShockTeleop
+
+        # instantiate the teleoperating device
+        ps4 = PS4DualShockTeleop()
+
+        # start teleloperating the robot
+        ps4.start_teleop()
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:270]: When running this, I got this error:

```
TypeError: PS4DualShockTeleop.__init__() missing 1 required positional argument: 'ee_frame_name'
```

As the error message says, this works, for instance

```
ps4 = PS4DualShockTeleop(ee_frame_name="panda_hand")
```

File: doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+        # plan to goal
+        plan_result = panda_arm.plan()
+
+        # execute the plan
+        if plan_result:
+                robot_trajectory = plan_result.trajectory
+                moveit.execute(robot_trajectory, blocking=True, controllers=[])
+
+We can perform motion planning interactively (see the motion planning tutorial for further details of the motion planning API).Suppose we are developing our code and we make a mistake such as follows: ::
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:206]: Missing space between sentences. Also can you add a link to the motion planning tutorial once merged?

```suggestion
We can perform motion planning interactively (see the motion planning tutorial for further details of the motion planning API). Suppose we are developing our code and we make a mistake such as follows: ::
```

File: doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst [hunk 5]
```diff
@@ -0,0 +1,294 @@
...
+
+        # execute the plan
+        if plan_result:
+                robot_trajectory = plan_result.trajectory
+                moveit.execute(robot_trajectory, blocking=True, controllers=[])
+
+Since we are using a notebook this mistake is easy to rectify without having to recompile any files. Simply edit the above notebook to match the below and rerun the cell: ::
+
+        # set plan start state using predefined state
+        panda_arm.set_start_state_to_current_state() # This conflicts with the current robot pose and will cause and error
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:231]: We want to remove this comment about the error in the "rectified" version, yeah?

File: doc/examples/jupyter_notebook_prototyping/launch/notebook_start_servo_client.launch.py [hunk 6]
```diff
@@ -0,0 +1,70 @@
...
+def generate_launch_description():
+    moveit_config = (
+        MoveItConfigsBuilder(
+            robot_name="panda", package_name="moveit_resources_panda_moveit_config"
+        )
+        .robot_description(file_path="config/panda.urdf.xacro")
+        .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
+        .moveit_cpp(
+            file_path=get_package_share_directory("moveit_py_example")
+            + "/config/moveit_py.yaml"
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/launch/notebook_start_servo_client.launch.py:36]: Similar to the other comments, use `os.path.join()` here and don't use slashes in your paths!

File: doc/examples/jupyter_notebook_prototyping/src/moveit_notebook_tutorial.ipynb [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+    "\n",
+    "Welcome to this tutorial on using jupyter notebooks with Moveit 2. A great benefit of being able to interact with MoveIt via a Python notebook is the ability to rapidly prototype code. We hope you find this interface intuitive and that you gain value from using MoveIt via Python notebooks.\n",
+    "\n",
+    "In this tutorial we will cover the following: \n",
+    "\n",
+    "* The required imports to run the notebook\n",
+    "* A motion planning example\n",
+    "* A teleoperation example\n",
+    "\n",
+    "If you have suggestions or feedback for this tutorial please post an issue on GitHub (https://github.com/ros-planning/moveit2_tutorials) and tag @peterdavidfagan."
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/src/moveit_notebook_tutorial.ipynb:18]: You sure you want to tell people to directly tag you in the issues? ;)

I personally wouldn't do this, haha.

**Review by sea-bass — CHANGES_REQUESTED** [review]
When testing I got this error message:

```
[ERROR] [launch]: Caught exception in launch (see debug for traceback): Caught exception when trying to load file of format [py]: "File /config/jupyter_notebook_prototyping.yaml doesn't exist"
```

And similar things with the RViz config. These were both solved by removing the `/` characters in your calls to `os.path.join()` and using it in the intended way. See my inline comments.

I also saw an error with the teleop section (though I don't have a dualshock), so see that too.

File: doc/examples/jupyter_notebook_prototyping/launch/jupyter_notebook_prototyping.launch.py (hunk 1 above)
Comment (peterdavidfagan) [reply doc/examples/jupyter_notebook_prototyping/launch/jupyter_notebook_prototyping.launch.py:23]: My apologies I rushed this change.

File: doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst (hunk 3 above)
Comment (peterdavidfagan) [reply doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:270]: I made a breaking change to the library which causes this error recently. I will update the notebook, ideally this should be pulled from parameters so I am adding an issue to update this.

Comment (peterdavidfagan) [reply doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:270]: https://github.com/ros-planning/moveit2/issues/1969

File: doc/examples/jupyter_notebook_prototyping/launch/notebook_start_servo_client.launch.py (hunk 6 above)
Comment (peterdavidfagan) [reply doc/examples/jupyter_notebook_prototyping/launch/notebook_start_servo_client.launch.py:36]: I was thinking of merging the launch files as I have done [here](https://github.com/peterdavidfagan/panda_moveit_demos/blob/main/notebook_demos/launch/moveit_py.launch.py). Do you agree with this design?

File: doc/examples/jupyter_notebook_prototyping/src/moveit_notebook_tutorial.ipynb (hunk 7 above)
Comment (peterdavidfagan) [reply doc/examples/jupyter_notebook_prototyping/src/moveit_notebook_tutorial.ipynb:18]: :) I am committed to fixing issues with the library so long as I can alongside my research projects. I am happy to have people ping me to make me aware of issues with the library.

File: doc/examples/jupyter_notebook_prototyping/launch/notebook_start_servo_client.launch.py (hunk 6 above)
Comment (sea-bass) [reply doc/examples/jupyter_notebook_prototyping/launch/notebook_start_servo_client.launch.py:36]: Yes, that sounds good!

File: doc/examples/jupyter_notebook_prototyping/src/moveit_notebook_tutorial.ipynb [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+    "# this is since we will start spinning a moveit_py node within this notebook.\n",
+    "\n",
+    "moveit_config = (\n",
+    "        MoveItConfigsBuilder(robot_name=\"panda\", package_name=\"moveit_resources_panda_moveit_config\")\n",
+    "        .robot_description(file_path=\"config/panda.urdf.xacro\")\n",
+    "        .trajectory_execution(file_path=\"config/gripper_moveit_controllers.yaml\")\n",
+    "        .moveit_cpp(\n",
+    "            file_path=get_package_share_directory(\"moveit2_tutorials\")\n",
+    "            + \"/config/jupyter_notebook_prototyping.yaml\"\n",
+    "        )\n",
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/src/moveit_notebook_tutorial.ipynb:66]: Update this comment to use os.path.join as well

File: doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+Define our MoveIt configuration, this step will also be important for later when configuring our notebook. ::
+
+        moveit_config = (
+                MoveItConfigsBuilder(robot_name="panda", package_name="moveit_resources_panda_moveit_config")
+                .robot_description(file_path="config/panda.urdf.xacro")
+                .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
+                .moveit_cpp(
+                        file_path=get_package_share_directory("moveit2_tutorials")
+                        + "/config/jupyter_notebook_prototyping.yaml"
+                )
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:84]: Update these docs to use os.path.join too

File: doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+* **rviz_node:** starts rviz2 for visualization purposes.
+* **static_tf:** publishes the static transforms between the world frame and panda base frame.
+* **robot_state_publisher:** publishes updated robot state information (transforms).
+* **ros2_control_node:** used to control groups of joints.
+
+::
+
+        rviz_config_file = (
+                get_package_share_directory("moveit2_tutorials") + "/config/jupyter_notebook_prototyping.rviz"
+        )
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:99]: Update these docs to use os.path.join too

File: doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+Once we have imported the required packages we need to define our ``moveit_py`` node configuration. We do this through using the ``MoveItConfigsBuilder`` as follows: ::
+
+        moveit_config = (
+                MoveItConfigsBuilder(robot_name="panda", package_name="moveit_resources_panda_moveit_config")
+                .robot_description(file_path="config/panda.urdf.xacro")
+                .trajectory_execution(file_path="config/gripper_moveit_controllers.yaml")
+                .moveit_cpp(
+                file_path=get_package_share_directory("moveit2_tutorials")
+                + "/config/jupyter_notebook_prototyping.yaml"
+                )
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:241]: Same here with os.path.join too

File: doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+        pose_goal.pose.position.z = 0.5
+        panda_arm.set_goal(goal_pose_msg = pose_goal, link_name = "panda_link8")
+
+        # plan to goal
+        plan_result = panda_arm.plan()
+
+        # execute the plan
+        if plan_result:
+                robot_trajectory = plan_result.trajectory
+                moveit.execute(robot_trajectory, blocking=True, controllers=[])
```
Comment (sea-bass) [inline doc/examples/jupyter_notebook_prototyping/jupyter_notebook_prototyping_tutorial.rst:289]: All the comments here need to match the updated code with the helper function and other changes

**Review by sea-bass — COMMENTED** [review]
Ensure the comments in the tutorial and the notebook match the changes to the launch files and the code -- also including the stuff that you recently changed in the motion planning API tutorial, like the helper function, the "else" condition when planning fails, etc.

After that, let me know when this is ready and I'll give it a last look!

**Comment by peterdavidfagan** [comment]
@sea-bass the above points were addressed. I can do a secondary review tomorrow morning in case you wish to avoid encountering any further review items. I think it is mostly good to go now.

**Review by sea-bass — APPROVED** [review]
Let's do this!

**Comment by peterdavidfagan** [comment]
🎉 🚀 🚢 🤖.

**Pull request merged**